### PR TITLE
DENG-2823: Added deprecate cli command

### PR DESCRIPTION
**Pull request by wwyc**
New cli command is added to support deprecation process.
`Deletion date` field is added to help keep track when tables are due to deletion.  This will allow us to capture deletion dates in labels and query them in `bigquery_tables_inventory`.

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-3098)

### CODE REVIEW

**Review by kwindau — APPROVED** [review]

**Review by Marlene-M-Hirose — APPROVED** [review]

**Comment by scholtzan** [comment]
It would be cool if the command could spit out a warning if it finds queries that are still referencing the table to be deprecated

File: bigquery_etl/cli/metadata.py [hunk 1]
```diff
@@ -126,3 +128,51 @@ def publish(name: str, sql_dir: Optional[str], project_id: Optional[str]) -> Non
...
+    "-d",
+    help="Date when table is scheduled for deletion. Date format: yyyy-mm-dd",
+    type=click.DateTime(formats=["%Y-%m-%d"]),
+    default=datetime.today() + relativedelta(months=+3),
+)
+def deprecate(
+    name: str,
+    sql_dir: Optional[str],
+    project_id: Optional[str],
+    deletion_date,
```
Comment (sean-rose) [inline bigquery_etl/cli/metadata.py:159]: Suggest type hint here for consistency.

File: bigquery_etl/cli/metadata.py [hunk 2]
```diff
@@ -126,3 +128,51 @@ def publish(name: str, sql_dir: Optional[str], project_id: Optional[str]) -> Non
...
+)
+@click.argument("name")
+@project_id_option(
+    ConfigLoader.get("default", "project", fallback="moz-fx-data-shared-prod")
+)
+@sql_dir_option
+@click.option(
+    "--deletion_date",
+    "--deletion-date",
+    "-d",
```
Comment (sean-rose) [inline bigquery_etl/cli/metadata.py:150]: For cases like this where it isn't a common parameter IMO it'd be preferable not to have a short option, which are more cryptic.

File: bigquery_etl/cli/metadata.py [hunk 3]
```diff
@@ -126,3 +128,51 @@ def publish(name: str, sql_dir: Optional[str], project_id: Optional[str]) -> Non
...
+        metadata.deprecated = True
+        metadata.deletion_date = deletion_date.date()
+
+        metadata.write(metadata_file)
+        click.echo(f"Updated {metadata_file} with deprecation.")
+
+    if not table_metadata_files:
+        raise FileNotFoundError(f"No metadata file(s) were found for: {name}")
+
+    return None
```
Comment (sean-rose) [inline bigquery_etl/cli/metadata.py:178]: Seems like the `return None` could be left off.  Or is Black requiring that for some reason?

File: bigquery_etl/metadata/parse_metadata.py [hunk 4]
```diff
@@ -153,6 +154,7 @@ class Metadata:
     references: Dict = attr.ib({})
     external_data: Optional[ExternalDataMetadata] = attr.ib(None)
     deprecated: bool = attr.ib(False)
+    deletion_date: date = attr.ib(None)
```
Comment (sean-rose) [inline bigquery_etl/metadata/parse_metadata.py:157]:
```suggestion
    deletion_date: Optional[date] = attr.ib(None)
```

File: bigquery_etl/cli/metadata.py (hunk 1 above)
Comment (BenWu) [reply bigquery_etl/cli/metadata.py:159]: I'll also add that `sql_dir` and `project_id` are optional as command line args but they're not expected to be null in the function.  I think `paths_matching_name_pattern` would actually error if sql_dir was null.

This might cause false warnings in code inspectors but maybe not in this case since `paths_matching_name_pattern` doesn't use type hints.

File: bigquery_etl/cli/metadata.py [hunk 5]
```diff
@@ -125,4 +127,49 @@ def publish(name: str, sql_dir: Optional[str], project_id: Optional[str]) -> Non
...
+        metadata.deprecated = True
+        metadata.deletion_date = deletion_date.date()
+
+        metadata.write(metadata_file)
+        click.echo(f"Updated {metadata_file} with deprecation.")
+
+    if not table_metadata_files:
+        raise FileNotFoundError(f"No metadata file(s) were found for: {name}")
+
     return None
```
Comment (sean-rose) [inline bigquery_etl/cli/metadata.py:175]: I just noticed this other `return None` that could be removed, and that actually this is currently resulting in `return None` being removed from the `publish()` function, which is unrelated to this PR.

So I'd suggest removing `return None` here and putting it back at the end of the `publish()` function to avoid having an unrelated change.

File: tests/cli/test_cli_metadata.py [hunk 6]
```diff
@@ -311,3 +315,73 @@ def test_metadata_publish_with_no_metadata_file(
...
+            distutils.dir_util.copy_tree(str(TEST_DIR), str(tmpdirname))
+
+            qualified_table_name = (
+                "moz-fx-data-shared-prod.telemetry_derived.clients_daily_v6"
+            )
+            result = runner.invoke(
+                deprecate,
+                [
+                    qualified_table_name,
+                    "--deletion_date=2024-03-02",
```
Comment (kik-kik) [inline tests/cli/test_cli_metadata.py:354]: Should we also have a test case for when the date format passed is incorrect?

**Review by kik-kik — APPROVED** [review]
